### ISSUE TITLE
[FR] Autologging functionality for scikit-learn integration with XGBoost (and LightGBM)

### ISSUE THREAD

**Issue by dalavayi** [mlflow/mlflow#4296]
## MLflow Roadmap Item

This is an MLflow Roadmap item that has been prioritized by the MLflow maintainers. We're seeking help with the implementation of roadmap items tagged with the `help wanted` label. Items tagged with the `needs design` label are awaiting a design sketch from an MLflow maintainer: @WeichenXu123.

For requirements clarifications and implementation questions, or to request a PR review, please tag @WeichenXu123 in your communications related to this issue.

## Proposal Summary

Hello, 

When trying to implement autolog() function for XGBoostClassifier and XGBoostRegressor, we are not able to see the metrics and parameters or models which are getting logged. Can you please address this issue. 

Thanks.

PFA added code for XGBoostRegressor which is to be converted to .py file to replicate the issue
[boston.txt](https://github.com/mlflow/mlflow/files/6386251/boston.txt)

**Comment by dalavayi** [comment]
Hi, Just wanted to check if the autologging functionality for XGboost SQL API is on the roadmap for future release?

**Comment by dbczumar** [comment]
@dalavayi Thanks for filing this! We're definitely interested in this capability. It would also be nice to support the LightGBM + scikit-learn integration: https://lightgbm.readthedocs.io/en/latest/Python-API.html#scikit-learn-api

**Comment by jwyyy** [comment]
Hi @dbczumar , is someone currently working on this issue? If not, I am willing to take the challenge and do the implementation. Can make a proposal first and work on the coding once it gets approved. Thanks!

**Comment by jwyyy** [comment]
Hi @dbczumar @WeichenXu123, just a follow-up on my last message. Can I start to work on this issue and make an initial plan for you to review? Thanks!

**Comment by dbczumar** [comment]
Hi @jwyyy , apologies for the delayed response. We're very excited about your contribution to this feature. Feel free to go ahead and make an initial plan. Thank you in advance for your contribution!

**Comment by jwyyy** [comment]
Thanks for your confirmation! I will start to make a plan shortly and notify you for a review before the actual implementation.

**Comment by jwyyy** [comment]
Hi @dbczumar @WeichenXu123, I have an update on the autologging functionality for XGBoost. The key is [L672](https://github.com/mlflow/mlflow/blob/master/mlflow/xgboost.py#L672). The `train()` of `xgboost` is reset to the newly defined `train()` inside `autolog()`. However, the original `train()` is defined in `xgboost.training` and imported when the package is initialized ([L10](https://github.com/dmlc/xgboost/blob/d8cb395380990cc627e6d208e7e9594a24cb8c2e/python-package/xgboost/__init__.py#L10)). On the other hand, `XGBoostClassifier` and `XGBoostRegressor` use [`train()`](https://github.com/dmlc/xgboost/blob/d8cb395380990cc627e6d208e7e9594a24cb8c2e/python-package/xgboost/sklearn.py#L787) imported from `training` ([L14](https://github.com/dmlc/xgboost/blob/d8cb395380990cc627e6d208e7e9594a24cb8c2e/python-package/xgboost/sklearn.py#L14)). The [`_wrap_patch()`](https://github.com/mlflow/mlflow/blob/efda4ca59ec1c37dd92fb6d70a88018783c0082b/mlflow/utils/autologging_utils/safety.py#L614) resets `xgboost.train()`but not `xgboost.training.train()`. An easy fix would be adding the following two lines in `mlflow.xgboost.autolog()`:
```
import xgboost.training
...
safe_patch(FLAVOR_NAME, xgboost.training, "train", train, manage_run=True)
```
(I have tested this. It will log metrics, params, and models.)

The autologging functionality issue for LightGBM is similar: please see [L563](https://github.com/mlflow/mlflow/blob/master/mlflow/lightgbm.py#L563), [L10](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/__init__.py#L10), [L15](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/sklearn.py#L15), and [L748](https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/sklearn.py#L748).

I will look into model logging/saving/loading more, since the class objects are different. (But it should be similar to the previous identified cause.)

**Comment by jwyyy** [comment]
@dbczumar Thank you for your feedback! The short answer to your question is only parameters passed to `train()` are logged, including **some** passed to `XGBoostClassifier.fit()`. Here are more details:
1.[ `params`](https://github.com/dmlc/xgboost/blob/d8cb395380990cc627e6d208e7e9594a24cb8c2e/python-package/xgboost/sklearn.py#L788) passed to `train()` includes parameters that are passed to constructor ([L776](https://github.com/dmlc/xgboost/blob/d8cb395380990cc627e6d208e7e9594a24cb8c2e/python-package/xgboost/sklearn.py#L776)). Parameters, such as `early_stopping_rounds`, `verbose`, and `callbacks`, in `XGBoostClassifier.fit()` are also logged, because they are ultimately passed to `train()`.
2. Other parameters passed to `XGBoostClassifier.fit()` are used to create data matrices ([L765-770](https://github.com/dmlc/xgboost/blob/d8cb395380990cc627e6d208e7e9594a24cb8c2e/python-package/xgboost/sklearn.py#L765-L770)). They are not logged in the current autologging for `xgboost.train()` [(L389)](https://github.com/mlflow/mlflow/blob/master/mlflow/xgboost.py#L389). My understanding is the current autologging for `xgboost.train()` only logs input data schema and model signature (see [L387](https://github.com/mlflow/mlflow/blob/master/mlflow/xgboost.py#L387) and [L635-660](https://github.com/mlflow/mlflow/blob/master/mlflow/xgboost.py#L635-L660)). 
3. It sort of makes sense, because often users use trained models to make predictions on unseen data. On the other hand, information like `sample_weight`, `base_margin`, and `feature_weights` can be helpful when tuning/debugging models, especially when users want to re-train a previously trained model. I think they can be logged by revising `DMatrix.__init__()`. A caveat is that loading them with model can lead to potential bugs, because the input data may not be the same unless it was stored and then is reloaded.

**Comment by harupy** [comment]
btw can we patch `XGBoostClassifier.fit` instead?

Pros
- We can reuse the sklearn's autologging routine.
- We can easily obtain a fitted estimator.

Cons
- `xgboost.train` and `XGBoostClassifier.fit` generate different results.

**Comment by jwyyy** [comment]
> I think we can get the best of both worlds by using sklearn autologging and implementing feature importance within sklearn autologging for the case where the estimator comes from XGBoost. Does that sound reasonable, @jwyyy @harupy ?

This sounds good to me. @dbczumar  Another question: does it mean that the autologging will be put in `mlflow.sklearn` not `mlflow.xgboost`? It can be a potential confusion, because XGBoost estimators are wrappers of `sklearn.base.BaseEstimator` and it will be called from `xgboost` not `sklearn`. That is, which should users use to initialize autologgging, `mlflow.xgboost.autolog()` or `mlflow.sklearn.autolog()`.

> Regarding the file / folder structure, the sklearn autologging definition is somewhat larger & more complex than the xgboost definition, so we split things into multiple utility modules within a top-level `mlflow.sklearn` module, hence the folder structure for sklearn while xgboost remains a file.

Thank you for your explanation! 😄

**Comment by dbczumar** [comment]
> > Autologging for XGBoost sklearn estimators should be initiated through `mlflow.xgboost.autolog`. We can add an internal (prefixed with `_`) autolog API in `mlflow.sklearn` with an `xgboost_estimators=True/False` param. If `xgboost_estimators==true`, we only patch XGBoost sklearn estimators and associated them with the XGBoost autologging integration via `safe_patch("xgboost", ...)`. Otherwise, we patch the sklearn estimators (and _not_ the xgboost estimators) via `safe_patch("sklearn", ...)`, which is the current behavior.
> > Let me know if this sounds good!
> 
> Yes, looks good to me! I will start to work on its implementation unless there are other suggestions. Thanks!

Thanks a bunch, @jwyyy ! Sounds good!

**Comment by jwyyy** [comment]
This PR https://github.com/mlflow/mlflow/pull/4883 will add XGBoost estimators to `mlflow.sklearn.autolog()` but without early stopping and feature importance. With this PR, we need to add these two logging functions inside `mlflow.xgboost.autolog()`. Of course, as discussed above, it should be initialized from `mlflow.xgboost.autolog()`.

**Comment by dbczumar** [comment]
> This PR #4883 will add XGBoost estimators to `mlflow.sklearn.autolog()` but without early stopping and feature importance. With this PR, we need to add these two logging functions inside `mlflow.xgboost.autolog()`. Of course, as discussed above, it should be initialized from `mlflow.xgboost.autolog()`.
> 
> Should we wait until this PR has been merged into the master branch? @dbczumar @harupy

#4883 doesn't seem directly related to this work. I think we can proceed with a separate PR for XGBoost.

**Comment by harupy** [comment]
+1 for starting with XGBoost!

**Comment by dbczumar** [comment]
Closing this out now that this support has been added for both XGBoost and LightGBM. Thank you so much for your contributions, @jwyyy !

**Comment by jwyyy** [comment]
@dbczumar I am really happy to contribute to the community :smile:  Thank you a lot for your review!
